rfd/0087--access-request-notification-routing.md

```diff
@@ -0,0 +1,111 @@
+---
+authors: Hugo Hervieux (hugo.hervieux@goteleport.com)
+state: draft
+---
+# RFD 87 - Access request notification routing
+
+## What
+
+Provide a granular way for administrators to configure which access-requests notifications should be delivered to whom over which medium.
+
+## Why
+
+- Teleport access request routing is not configured the same way depending on the plugin: Pagerduty uses annotations while most plugins use a role_to_recipient map in the plugin config. Users leveraging both PagerDuty and Slack plugins have to deal with two configuration languages.
+- Current Pagerduty request routing applies the same rule to all requests from the same role, regardless of the requested role. All access requests don't have the same severity, users wanting to route access requests differently have to create multiple roles. Depending on the company size and structure cardinality can become a problem. See https://github.com/gravitational/teleport-plugins/issues/597
+- Access request routing map baked in the plugin deployment:
+  - Will become huge for big orgs
+  - Requires redeploying each time a change is done
+  - Does not allow to route requests regarding of the requestor
+
+## Details
+
+### User story
+
+Bob is a developer. He deploys code using CI/CD pipelines. He can request access to prod in read-only mode and dev in read-write mode for debugging. In case of incident he can request prod read-write access.
+Alice is the lead-dev, she grants regular access requests through MsTeams during open hours and can also approve urgent read-write requests.
+Alice should not be paged each time Bob needs to debug something during open hours, but in case of incident Bob needs immediate access and Alice should be paged.
+
+In Teleport terms:
+
+- the role `developer` can request roles `dev-rw`, `prod-ro` and `prod-rw`
+- the role `lead-developer` can accept requests
+- only `prod-rw` access requests should trigger a PagerDuty incident
+- `dev-rw` and `prod-ro` access requests should trigger a MsTeams message
+
+### Suggestion 1: With a where clause
+
+```
+kind: role
+metadata:
+  name: developer
+spec:
+  allow:
+    request:
+      roles: ["dev-rw", "prod-ro","prod-rw"]
+      destinations:
```

Comment:
I think this looks better than annotations. I think we should pursue similar route (see below for details).

As far as I understand, having `destinations` will impact how access request is annotated, which in turn will impact how plugins will evaluate the requests. 

For example, this rule

```
      - where: 'equals(resource.spec.requested_role, "prod-rw")'
        plugin: msteams
        target: ["Alice@example.com"]
```
Will result in access request with this annotation if the role is `prod-rw`:

```
request:
   annotations:
      plugin: msteams
      notify: ["alice@example.com"]
```

MsTeams plugin will process this access request, while PagerDuty will ignore it. 

**Separate resource**

Instead of adding more logic to roles, I would like to introduce a resource similar to `login rules`.

Please review [this RFD](https://github.com/gravitational/teleport/pull/14977/files) proposal.

I would like us to introduce a separate `notification_rule` resource that will have a similar structure:

```
kind: access_routing_rule
version: v1
metadata:
  name: example
spec:
# same priority will be ordered by a lexicographical sort by their names.
  priority: 0
  # targets returns a list of (plugin, set(target)) 
  targets: >
    choose(
        option(
             equals(resource.spec.requested_role, "prod-rw"), pair("pagerduty,  set("Teleport Alice")))
      option(
             equals(resource.spec.requested_resources, ...
)
```

We also should investigate how this would work with JIT resource based access requests as well.

Reply:
First thoughts:
- moving the rules out of the role will avoid pollution of the role spec, it also supports interesting many-to-many bindings between `roles` and `access_routing_rules`
- Before emitting the access request Teleport has to iterate over all `access_routing_rules` as it won't be able to know if an `access_routing_rule` could be relevant for this access request. Processing time and load are O(n). How many `access_routing_rules` a big corp is going to create? How to deal with 1k or 10k rules?
- With multiple teams maintaining their own servers/roles in Teleport they will write rules potentially conflicting. This is not something new, conflict could already happen and allow users from teamA to have access to some servers from teamB. However, a misconfigured rule with low priority will act as a back hole and potentially lock users out.
- The predicate language can be complex for a first time user, I think we can lower the overhead by removing the need for the first `choose(option(),option(),...` in the general case, while still supporting everything in a string if they want/need something more complex (no condition evaluates to true).

```yaml
kind: access_routing_rule
version: v1
metadata:
  name: example
spec:
  priority: 0
  targets:
  # regular use-case
  - condition: 'equals(resource.spec.requested_role, "prod-rw")'
    value: 'pair("pagerduty",  set("Teleport Alice"))'
  # if for some reason the user wants to apply custom logic
  - value: >
      choose(
        option(
          equals(...), choose(option(),option()),
        option(....))
```

This layout also looks more templatable for power users who automate their Teleport configuration.

- We could go further with the `access_routing_rule` by allowing to set any arbitrary annotation rather than `plugin` and `notify` (returning a dict of annotations instead of a pair). I'm not sure if there's a use-case though.
- Introducing the ressource will add a degree of freedom, we might want to investigate different models like "one rule per notification recipient", "one rule per requested role", "one rule per requestor" and present them to the users with pros/cons.

Reply:
I made a v2 based on your suggestions, the main changes with to what you suggested is the ability for a single request to support multiple targets. This solves the blackhole issue and allows simpler backward compatibility. It also simplifies a bit what users have to input and the amount of predicate language they have to understand before using the feature.

I also added security and performance configurations sections. Can you review it again and see if this is closer to what you expected?